### DISCUSSION TITLE
3d support

### DISCUSSION THREAD

**Discussion by ecomodeller** [Deltares/xugrid#242] [General]
Is there any support at the moment in `xugrid` for using a [UGRID 3d layered topology](https://ugrid-conventions.github.io/ugrid-conventions/#3d-layered-mesh-topology), or planned?

**Comment by Huite** [comment]
Hi @ecomodeller,

3D layered topologies should generally work out of the box, since horizontal and vertical dimensions are treated as fully separately dimensions. Since the vertical (layer) dimension is fully orthogonal, (nearly) all operations are possible using standard xarray functionality. In most cases, the layer coordinates would be stored with dimensions (layer, face) or (layer, node) when they vary in the horizontal plane or just (layer,) when they are constant in the horizontal plane.

We use the layered topology for our groundwater modelling with MODFLOW 6, see e.g.: https://deltares.github.io/imod-python/examples/mf6/circle_transport.html

To illustrate, at the end it has this example:

```python
concentration.isel(time=-1).ugrid.sel(y=0).plot.contourf(
    ax=ax, x="mesh2d_x", y="z", cmap="RdYlBu_r"
)
```

Which generates a vertical cross section plot. It uses standard xarray `isel` to select over the time dimensions, then a `ugrid.sel` for the horizontal part. The result is an ordinary xarray DataArray which can be plotted with `.plot.contourf`.

There are some specific things which do require both horizontal and vertical coordinates at the same time. The primary one I have in mind is regridding from one layered model to another layered model, where the layers are spaced differently -- this is a common operation for groundwater modeling at least.

Anyway, you should be able to create and open such datasets. If you run into trouble, I'd be happy to hear what's going wrong.
